company-template.el

```diff
@@ -55,19 +60,51 @@
 
 (defun company-template-forward-field ()
   (interactive)
+  (let ((start (point))
+        (next-field-start (company-template-find-next-field)))
+    (push-mark)
+    (goto-char next-field-start)
+    (company-template-remove-field (company-template-field-at start))))
+
+(defun company-template-clear-field ()
+  "Clear the field at point."
+  (interactive)
+  (let ((ovl (company-template-field-at (point))))
+    (when ovl
+      (company-template-remove-field ovl t)
+      (let ((after-clear-fn (overlay-get ovl 'after-clear-function)))
```

Comment:
`delete-overlay` doesn't "kill" the object, it just makes it point to no buffer, see the doc of `overlay-buffer` for confirmation.

The overlay will be GCd eventually, so it  should be fine.